cmd/dep/gb_importer.go

```diff
@@ -0,0 +1,205 @@
+		// multiple entries for the same project root, one for each imported subpackage
+		var root gps.ProjectRoot
+		var err error
+		if root, err = i.sm.DeduceProjectRoot(pkg.Importpath); err != nil {
+			return nil, nil, err
+		}
+
+		// Set the proper import path back on the dependency
+		pkg.Importpath = string(root)
+
+		// If we've already locked this project root then we can skip
+		if projectExistsInLock(lock, pkg.Importpath) {
+			continue
+		}
+
+		// Otherwise, attempt to convert this specific package, which returns a constraint and a lock
+		pc, lp, err := i.convertOne(pkg)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		manifest.Constraints[pc.Ident.ProjectRoot] = gps.ProjectProperties{Source: pc.Ident.Source, Constraint: pc.Constraint}
+		lock.P = append(lock.P, lp)
+
+	}
+
+	return manifest, lock, nil
+}
+
+func (i *gbImporter) convertOne(pkg gbDependency) (pc gps.ProjectConstraint, lp gps.LockedProject, err error) {
+	/*
+		gb's vendor plugin (gb vendor), which manages the vendor tree and manifest
+		file, supports fetching by a specific tag or revision, but if you specify
+		either of those it's a detached checkout and the "branch" field is HEAD.
+		The only time the "branch" field is not "HEAD" is if you do not specify a
+		tag or revision, otherwise it's either "master" or the value of the -branch
+		flag
+
+		This means that, generally, the only possible "constraint" we can really specify is
+		the branch name if the branch name is not HEAD. Otherwise, it's a specific revision.
+
+		However, if we can infer a tag that points to the revision or the branch, we may be able
+		to use that as the constraint
+
+		So, the order of operations to convert a single dependency in a gb manifest is:
+			- Find a specific version for the revision (and branch, if set)
+			- If there's a branch available, use that as the constraint
+			- If there's no branch, but we found a version from step 1, use the version as the constraint
+	*/
+	pc.Ident = gps.ProjectIdentifier{ProjectRoot: gps.ProjectRoot(pkg.Importpath), Source: pkg.Repository}
+
+	// Generally, gb tracks revisions
+	var revision = gps.Revision(pkg.Revision)
+
+	// But if the branch field is not "HEAD", we can use that as the initial constraint
+	var constraint gps.Constraint
+	if pkg.Branch != "" && pkg.Branch != "HEAD" {
+		constraint = gps.NewBranch(pkg.Branch)
+	}
+
+	// See if we can get a version from that constraint
+	version, err := lookupVersionForLockedProject(pc.Ident, constraint, revision, i.sm)
+	if err != nil {
+		// Log the error, but don't fail it. It's okay if we can't find a version
+		i.logger.Println(err.Error())
+	}
+
+	// If the constraint is nil (no branch), but there's a version, infer a constraint from there
+	if constraint == nil && version != nil {
+		constraint, err = i.sm.InferConstraint(version.String(), pc.Ident)
+		if err != nil {
+			return
+		}
+	}
+
+	// If there's *still* no constraint, set the constraint to the revision
```

Comment:
Is this the proper behavior? Should there just not be a constraint if there's no branch specified and no tag that points to the revision?